e2e/stack.go

```diff
@@ -121,74 +129,80 @@ func (s *Stack) Run(ctx context.Context, env *environment.Env) error {
 			Address: &userAddress,
 		}
 		l1state.OnAccount(user.Address, user)
-		s.L1Users = append(s.L1Users, L1User{
+		l1users = append(l1users, L1User{
 			Address:    userAddress,
 			PrivateKey: privateKey,
 		})
 	}
 
 	l1genesis, err := opgenesis.BuildL1DeveloperGenesis(deployConfig, l1state, l1Deployments)
 	if err != nil {
-		return fmt.Errorf("build l1 developer genesis: %v", err)
+		return nil, fmt.Errorf("build l1 developer genesis: %v", err)
 	}
 
 	l1client, l1HTTPendpoint, err := gethdevnet(s.l1BlockTime, l1genesis)
 	if err != nil {
-		return fmt.Errorf("ethdevnet: %v", err)
+		return nil, fmt.Errorf("ethdevnet: %v", err)
 	}
 
 	l1url, err := url.ParseString(l1HTTPendpoint)
 	if err != nil {
-		return fmt.Errorf("new l1 url: %v", err)
+		return nil, fmt.Errorf("new l1 url: %v", err)
 	}
 
 	// NOTE: should we set a timeout on the context? Might not be worth the complexity.
 	if !l1url.IsReachable(ctx) {
-		return fmt.Errorf("l1 url not reachable: %s", l1url.String())
+		return nil, fmt.Errorf("l1 url not reachable: %s", l1url.String())
 	}
 
 	l1 := NewL1Client(l1client)
 
 	latestL1Block, err := l1.BlockByNumber(ctx, nil)
 	if err != nil {
-		return fmt.Errorf("get the latest l1 block: %v", err)
+		return nil, fmt.Errorf("get the latest l1 block: %v", err)
 	}
 
 	// Run Monomer.
 	if err := s.runMonomer(ctx, env, latestL1Block.Time(), deployConfig.L2ChainID); err != nil {
-		return err
+		return nil, err
 	}
 	if !s.monomerEngineURL.IsReachable(ctx) {
-		return nil
+		return nil, fmt.Errorf("reaching monomer url: %s", s.monomerEngineURL.String())
 	}
 	monomerRPCClient, err := rpc.DialContext(ctx, s.monomerEngineURL.String())
 	if err != nil {
-		return fmt.Errorf("dial monomer: %v", err)
+		return nil, fmt.Errorf("dial monomer: %v", err)
 	}
 	monomerClient := NewMonomerClient(monomerRPCClient)
 	l2GenesisBlockHash, err := monomerClient.GenesisHash(ctx)
 	if err != nil {
-		return fmt.Errorf("get Monomer genesis block hash: %v", err)
+		return nil, fmt.Errorf("get Monomer genesis block hash: %v", err)
 	}
 
 	rollupConfig, err := deployConfig.RollupConfig(latestL1Block, l2GenesisBlockHash, 1)
 	if err != nil {
-		return fmt.Errorf("new rollup config: %v", err)
+		return nil, fmt.Errorf("new rollup config: %v", err)
 	}
 
 	opStack := NewOPStack(
 		l1url,
 		s.monomerEngineURL,
 		s.opNodeURL,
 		l1Deployments.L2OutputOracleProxy,
-		s.L1Users[0].PrivateKey,
+		l1users[0].PrivateKey,
 		rollupConfig,
 		s.eventListener,
 	)
 	if err := opStack.Run(ctx, env); err != nil {
-		return fmt.Errorf("run the op stack: %v", err)
+		return nil, fmt.Errorf("run the op stack: %v", err)
 	}
-	return nil
+
+	return &StackConfig{
+		L1URL:    l1url,
+		Config:   rollupConfig,
+		Operator: l1users[0],
+		Users:    l1users[1:],
+	}, nil
```

Comment:
I haven't looked through the PR in detail yet, but my impression is that this can be quite dangerous. The caller will be able to modify the data provided in this `StackConfig`. Probably better for the caller to pass in this kind of information to the `Stack`.

Reply:
On second thought, my points don't make much sense. I'll let Nate take the reviews here.

Reply:
Third thought - I think these points do make sense. The returned data should be "information only", and not knobs of the running stack that can be twiddled by the caller.

Tracked in [notion](https://www.notion.so/nethermind/e2e-return-copied-data-from-stack-Run-db0e763ab0d8489598574f718e3d8edd?pvs=4).